examples/amm/src/lib.rs

```diff
@@ -67,23 +67,48 @@ pub enum ApplicationCall {
 }
 
 pub fn calculate_output_amount(
-    input_amount: u64,
-    input_pool_balance: u64,
-    output_pool_balance: u64,
-) -> Result<u64, AmmError> {
-    if input_pool_balance == 0 || output_pool_balance == 0 {
+    input_amount: Amount,
+    input_pool_balance: Amount,
+    output_pool_balance: Amount,
+) -> Result<Amount, AmmError> {
+    if input_pool_balance == Amount::ZERO || output_pool_balance == Amount::ZERO {
         return Err(AmmError::InvalidPoolBalanceError);
     }
 
-    let numerator = input_amount * output_pool_balance;
-    let denominator = input_pool_balance + input_amount;
-
-    if denominator == 0 {
+    let input_amount_bigint = BigUint::from_u128(u128::from(input_amount))
+        .expect("Couldn't generate input_amount in bigint");
+    let output_pool_balance_bigint = BigUint::from_u128(u128::from(output_pool_balance))
+        .expect("Couldn't generate output_pool_balance in bigint");
+    let input_pool_balance_bigint = BigUint::from_u128(u128::from(input_pool_balance))
+        .expect("Couldn't generate input_pool_balance in bigint");
+
+    // Logic for this is the following:
+    // This is a Constant Product Automated Market Maker, or CPAMM, so we want
+    // the product to remain constant.
+    // That means that this is the equation we need to solve to find output_amount:
+    //      (input_pool_balance + input_amount) * (output_pool_balance - output_amount) = input_pool_balance * output_pool_balance
+    //      output_pool_balance - output_amount = (input_pool_balance * output_pool_balance) / (input_pool_balance + input_amount)
+    //      output_amount = output_pool_balance - (input_pool_balance * output_pool_balance) / (input_pool_balance + input_amount)
+    //      output_amount = (output_pool_balance * (input_pool_balance + input_amount) - (input_pool_balance * output_pool_balance)) / (input_pool_balance + input_amount)
+    //      output_amount = (input_pool_balance * output_pool_balance + input_amount * output_pool_balance - input_pool_balance * output_pool_balance) / (input_pool_balance + input_amount)
+    //      output_amount = (input_amount * output_pool_balance) / (input_pool_balance + input_amount)
+
+    // Numerator will be a number with 36 decimal points here
+    let numerator_bigint = input_amount_bigint.clone() * output_pool_balance_bigint;
+    // Denominator will have 18 decimal points
+    let denominator_bigint = input_pool_balance_bigint + input_amount_bigint;
+
+    if denominator_bigint.is_zero() {
         return Err(AmmError::DivisionByZero);
     }
 
-    let output_amount = numerator / denominator;
-
+    // Diving 36 decimal points with 18 decimal points = 18 decimal points
```

Comment:
:ocean: :diving_mask: :ocean: 

```suggestion
    // Dividing 36 decimal points with 18 decimal points = 18 decimal points
```

Reply:
🤣🤣 Nice catch!